src/Sylius/Bundle/CoreBundle/EventListener/TaxonDeletionListener.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\EventListener;
+
+use Doctrine\ORM\EntityManagerInterface;
+use Sylius\Component\Core\Model\TaxonInterface;
+use Sylius\Component\Core\Promotion\Checker\Rule\HasTaxonRuleChecker;
+use Sylius\Component\Core\Promotion\Checker\Rule\TotalOfItemsFromTaxonRuleChecker;
+use Sylius\Component\Promotion\Model\PromotionRuleInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\EventDispatcher\GenericEvent;
+use Webmozart\Assert\Assert;
+
+final class TaxonDeletionListener
+{
+    /** @var RepositoryInterface */
+    private $promotionRuleRepository;
+
+    /** @var EntityManagerInterface */
+    private $manager;
+
+    public function __construct(RepositoryInterface $promotionRuleRepository, EntityManagerInterface $manager)
+    {
+        $this->promotionRuleRepository = $promotionRuleRepository;
+        $this->manager = $manager;
+    }
+
+    public function removeTaxonFromPromotionRules(GenericEvent $event): void
```

Comment:
I have two problems with this implementation. First of all, it will work only for taxons deleted through Resource Controller. Secondly, we are hardcoded to only two existing rules that contain relation to the taxon(or taxons).

I would prefer, to decouple this logic from ResourceController, but I'm not sure if it's a good place to do it. For sure, it is a good place to start a discussion.
According to the second problem, it could be resolved with some new interface and iteration over all promotion rules in our store. Something like: TaxonsAwarePromotionRule or something similar. WDYT?

Reply:
We've discussed the first problem, that this behaviour could be associated with the resource event.
About the second one, I've moved the logic of removing taxon from configuration to separate services.